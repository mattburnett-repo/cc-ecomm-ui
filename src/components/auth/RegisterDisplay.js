
import { useContext } from 'react'
import { RegisterMessageContext } from "../../features/auth/register";


export default function Register( props ) {
    const { handleClick, handleRegister } = props.handlers
    const message = useContext(RegisterMessageContext)

    return (
        <div role="presentation" aria-label="register-display">
            <div role="presentation" aria-label="oauth">
                {/* FIXME: resolve href / onClick */}
                <div><a href="/auth/google" aria-label="google-auth" onClick={handleClick}>Log in with Google</a></div>
                <div><a href="/auth/github" aria-label="github-auth" onClick={handleClick}>Log in with GitHub</a></div>   
            </div>
            <div role="presentation" aria-label="register-display-form">
                {(message && <div role="presentation" aria-label='register-error-message'>{message}</div>)}
                <form onSubmit={handleRegister} method="post">
                    <div>
                        <label htmlFor="username">Username:</label>
                        <input id="username" aria-label="username" />
                    </div>
                    <div>
                        <label htmlFor="email">Email:</label>
                        <input id="email" type="email" aria-label="email" />
                    </div>
                    <div>
                        <label htmlFor="password">Password:</label>
                        <input id="password" type="password" aria-label="password" />
                    </div>
                    <div>
                        <label htmlFor="password2">Re-enter password:</label>
                        <input id="password2" type="password" aria-label="password" />
                    </div>
                    <div>
                        <button type="susbmit" aria-label="register-button">Create Account</button>
                    </div>
                </form>
            </div> 
            <div role="presentation" aria-label="login">
                <label htmlFor="login">Already have an account? </label>
                <a href='/login' id="login" aria-label="login-link">Login</a>
            </div>
        </div>
    )
}