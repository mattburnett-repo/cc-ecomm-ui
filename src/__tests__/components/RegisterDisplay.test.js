// 20211217: Wireframe https://wireframe.cc/6MXHyR

import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

import { ThemeProvider } from 'styled-components'
import theme from '../../components/styles/Theme'
// <ThemeProvider theme= { theme }> </ThemeProvider>

import RegisterDisplay from '../../components/auth/RegisterDisplay';

const mockHandleClick = jest.fn();
const mockHandleRegister = jest.fn()
const mockHandleLogin = jest.fn()

function mockHandlers() {
    return {
        handleClick: mockHandleClick,
        handleRegister: mockHandleRegister,
        handleLogin: mockHandleLogin,
    }
}

describe('<RegisterDisplay /> component tests', () => {
    beforeEach(() => render(<ThemeProvider theme= { theme }><RegisterDisplay handlers={mockHandlers()}/></ThemeProvider>));

    it('renders the display component', () => {
        screen.getByRole('presentation', { name: /^register-display$/i })
    })

    // it('renders an oauth section', () => {
    //     screen.getByRole('presentation', { name: /^oauth$/i })
    // })
    // it('renders a google oauth link', () => {
    //     let theVal = screen.getByRole('link', { name: /google-auth/i});
    //     expect(theVal).toHaveTextContent('Log in with Google');
    // });
    // it('clicks the Google auth link', () => {       
    //     // mockHandleClick.mockImplementation((e) => {console.log(e.target.text)});
    //     mockHandleClick.mockReturnValue('Log in with Google'); // FIXME: evaluate e.target.name / value instead of hard-coding test val here
    //     const theVal = screen.getByRole('link', {name: /google-auth/i});
    //     fireEvent.click(theVal);

    //     expect(mockHandleClick.mock.calls.length).toBe(1);
    //     expect(mockHandleClick()).toBe('Log in with Google');
    // });
    // it('renders a github oauth link', () => {
    //     let theVal = screen.getByRole('link', { name: /github-auth/i});
    //     expect(theVal).toHaveTextContent('Log in with GitHub');
    // });
    // it('clicks the GitHub auth link', () => {
    //     // mockHandleClick.mockImplementation((e) => {console.log(e.target.text)});
    //     mockHandleClick.mockReturnValue('Log in with GitHub'); // FIXME: evaluate e.target.name / value instead of hard-coding test val here
    //     const theVal = screen.getByRole('link', {name: /github-auth/i});
    //     fireEvent.click(theVal);

    //     expect(mockHandleClick.mock.calls.length).toBe(1);
    //     expect(mockHandleClick()).toBe('Log in with GitHub');               // FIXME
    // });

    it('renders a registration form', () => {
        screen.getByRole('presentation', { name: /register-display-form/i})
    })
    it('renders a username input', () => {
        screen.getByRole('textbox', { name: /username/i});
    });
    it('accept a value for username', ()=> {
        const theElement = screen.getByRole('textbox', {name: /username/i});
        userEvent.type(theElement, 'asdf');
        
        expect(theElement).toHaveValue('asdf');
    });
    it('renders an email input and accepts input', () => {
        const theElement = screen.getByRole('textbox', { name: /email/i});
        userEvent.type(theElement, 'asdf@asdf.com');
        
        expect(theElement).toHaveValue('asdf@asdf.com');
    });

    // I don't like using test ids, but can't find a simple / clean solution to test passwords,
    //      so here we are...
    it('renders a password input', () => {
        screen.getByTestId('password')
        // screen.getByLabelText('Password:');
    });
    it('accepts a value for password', ()=> {
        const theElement = screen.getByTestId('password')
        userEvent.type(theElement, 'asdf');
        
        expect(theElement).toHaveValue('asdf');
    });
    it('renders a password verify input', () => {
        screen.getByTestId('password-verify')
        // screen.getByLabelText('Password:');
    });
    it('accepts a value for password verify', ()=> {
        const theElement = screen.getByTestId('password-verify')
        userEvent.type(theElement, 'asdf');
        
        expect(theElement).toHaveValue('asdf');
    });

    it('renders a Register button', () => {
        screen.getByRole('button', { name: /register/i});
    });
    it('clicks the Register button', () => {
        mockHandleRegister.mockImplementation((e) => {e.preventDefault()});
        const theVal = screen.getByRole('button', {name: /register-button/i});
        fireEvent.click(theVal);

        expect(mockHandleRegister.mock.calls.length).toBe(1);
        // expect(mockHandleClick()).toBe('Log in'); 
    });

    it('renders a Login section', () => {
        let theVal = screen.getByRole('presentation', { name: /^login$/i});
        expect(theVal).toHaveTextContent('Login');
    });
    it('renders a Login link', () => {
        let theVal = screen.getByRole('link', { name: /login-link/i});
        expect(theVal).toHaveTextContent('Login');
    });
    it('clicks the Login link', () => {
        const theVal = screen.getByRole('link', {name: /login-link/i});
        fireEvent.click(theVal);

        // expect(theVal).toHaveBeenClicked() // FIXME: find the appropriate matcher
    });

    // it('renders a snapshot', () => {
    //     const tree = renderer.create(<ThemeProvider theme= { theme }><RegisterDisplay handlers={mockHandlers()}/></ThemeProvider>).toJSON();
    //     expect(tree).toMatchSnapshot();
    // });
}); // end component

