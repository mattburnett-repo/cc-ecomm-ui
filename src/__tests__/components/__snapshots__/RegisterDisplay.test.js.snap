// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterDisplay /> renders a snapshot renders a snapshot 1`] = `
<div>
  <div>
    <button
      aria-label="google-auth"
    >
      Log in with Google
    </button>
    <button
      aria-label="github-auth"
    >
      Log in with Google
    </button>
  </div>
  <div>
    <label
      htmlFor="username"
    >
      Username:
    </label>
    <input
      aria-label="username"
      id="username"
    />
  </div>
  <div>
    <label
      htmlFor="email"
    >
      Email:
    </label>
    <input
      aria-label="email"
      id="email"
      type="email"
    />
  </div>
  <div>
    <label
      htmlFor="password"
    >
      Password:
    </label>
    <input
      aria-label="password"
      id="password"
      type="password"
    />
  </div>
  <div>
    <button
      aria-label="register"
    >
      Create Account
    </button>
  </div>
  <div>
    <label
      htmlFor="login"
    >
      Already have an account? 
    </label>
    <a
      aria-label="login"
      href="/login"
      id="login"
    >
      Login
    </a>
  </div>
</div>
`;
